text/1022-freetypes.rst

```diff
@@ -0,0 +1,213 @@
+RFC 1022: Typing free objects & simplifying SDL syntax
+======================================================
+
+Abstract
+========
+
+This RFC proposes enhancements to the EdgeDB type system allowing users to
+explicitly type free objects. Free object types can then be used to type
+function parameters or used to indicate types of query arguments.
+
+While discussing the necessary new syntax for typing free objects, this
+RFC proposes a set of adjustments to the current EdgeDB Schema Declaration
+Language syntax.
+
+
+Typing free objects
+===================
+
+The proposed grammar is structured with the following requirements in mind:
+
+* The syntax must be "light" as it will appear frequently in EdgeQL
+  queries inside cast (e.g. ``<type_expr>$foo``) constructs.
+
+* We want to automatically infer as many things as possible.
+
+* We want the syntax to play nice with the existing syntax for typing
+  arrays and tuples.
+
+The grammar looks as follows::
+
+    type_name:
+        # name of the type, such as "int64" or "default::User"
+
+    type_expr:
+        # type expressions valid in SDL, such as "A | B" or
+        # "array<int64>"
+
+    property_or_link:
+        ["multi" | "single"] ["required" | "optional"] ["property" | "link"] \
+            name ":" type_name | type_expr | free_object_type
+
+    free_object_type:
+        "{"
+            property_or_link
+            [ "," property_or_link ] *
+        "}"
+
+
+Key takeaways:
+
+* EdgeDB will infer whether a field is a property or a link, making
+  "property" and "link" keywords optional.
+
+* Similar to SDL, all properties and links are *optional* and
+  *single* by default.
+
+(Note that the syntax deviates from SDL: we use ``:`` instead of ``->`` and
+``property`` / ``link`` modifier keywords are optional. To make the new free
+object types syntax compatible with SDL we also propose to update the SDL
+syntax in the below `Updates to SDL`_ section.)
+
+Examples:
+
+A free object as a query argument::
+
+    with
+        data := <{a: int64, b: str}>$arg
+    select
+        data.a + <str>data.b;
+
+A more complex free object with nested data and JSON casts::
+
+    with
+        data := <{
+            required pokemon: str,
+            multi required locations: {
+                required point: tuple<float64, float64>,
+                street: str
+            }
+        }>json_array_unpack(<json>$arg)
+
+    for item in data union (
+        insert Pokemon {
+            name := item.pokemon,
+            # ...
+        }
+    )
+
+In user-defined functions::
+
+    function compute(pokemon: {
+        required pokemon: str,
+        multi required locations: {
+            required point: tuple<float64, float64>,
+            street: str
+        }
+    }) -> int64 using (
+        select # ...
+    )
+
+
+Updates to SDL
+==============
+
+Summary:
+
+* Make the ``property`` and ``link`` keyword optional. The kind of a pointer
```

Comment:
Unfortunately I don't think we can drop it in *all* places, since for computeds it will conflict with existing `SetField` commands. (For things like `default := "foo"`.)

I think we can drop it for non-computed uses, though I'm still fighting with the parser generator a bit

Reply:
@msullivan Sully, since it looks like you made it work (with keeping "link" and "property" optional keywords) can you list all cases where link/property become optional and where they are needed to disambiguate?

Reply:
(I'll update the RFC with that)

Reply:
> Unfortunately I don't think we can drop it in all places, since for computeds it will conflict with existing SetField commands. (For things like default := "foo".)

That's only relevant for computable link properties though, right?

Reply:
That said requiring "link" / "property" modifiers for all computeds is OK-ish compromise, I think.

Reply:
Yeah, the one case where we can't drop from a syntax perspective is `foo := ...`. This is kind of a bummer, but oh well.

Syntactically, we definitely can do `multi foo := ...`, but I think it might be worth *not* allowing that:
1. For consistency, it might be best to always just require `link`/`property` on computeds, instead of just allowing it to be dropped if there is a cardinality annotation
2. There are implementation reasons: for non-computed links, we should be able to resolve whether they are links or properties in `declarative.py`, when we convert SDL to DDL, and so wouldn't ever need to handle DDL where it is not known. For computeds it is trickier, since we would need to actually typecheck the expression. Handling DDL with unknown pointer kind is probably possible but probably kind of invasive.